packages/firestore/test/integration/api/server_timestamp.test.ts

```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return previous value', () => {
+    let previousSnapshot: firestore.DocumentSnapshot;
+
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          previousSnapshot = snapshot;
+          verifyTimestampsAreResolved(snapshot);
+        })
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot =>
+          verifyTimestampsUsePreviousValue(snapshot, previousSnapshot)
+        )
+        .then(waitForRemoteEvent)
+        .then(snapshot => verifyTimestampsAreResolved(snapshot));
+    });
+  });
+
+  it('can return previous value of different type', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          verifyTimestampsAreResolved(snapshot);
+        })
+        .then(() =>
+          docRef.update('a', firebase.firestore.FieldValue.serverTimestamp())
+        )
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          expect(snapshot.get('a', { serverTimestamps: 'none' })).to.equal(
+            null
+          );
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
+          expect(snapshot.get('a', { serverTimestamps: 'previous' })).to.equal(
+            42
+          );
+        })
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          expect(
+            snapshot.get('a', { serverTimestamps: 'none' })
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'previous' })
+          ).to.be.instanceof(Date);
+        });
+    });
+  });
+
+  it('can return previous value through consecutive updates', () => {
+    return withTestSetup(() => {
+      // TODO(mikelehen): Find better way to expose this to tests.
+      // tslint:disable-next-line:no-any enableNetwork isn't exposed via d.ts
+      const firestoreInternal = docRef.firestore.INTERNAL as any;
+
+      let promises: Promise<void>[] = [];
+
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          verifyTimestampsUsePreviousValue(snapshot, null);
+          return waitForRemoteEvent();
+        })
```

Comment:
Can you see if you can clean up the setup here too?

Reply:
Done